extras/reset_mbients.py

```diff
@@ -1,77 +1,309 @@
+        :param connect_attempts: The number of times to try to connect to the device before giving up.
+        :param reset_timeout: How long to wait for the device to reset before giving up.
+        """
+        super().__init__()
+        self.device_info = device_info
+        self.connect_attempts = connect_attempts
+        self.reset_timeout = reset_timeout
+        self.logger = logging.getLogger('default')
+        self.disconnect_event = threading.Event()
+        self.success = False
+
+    def format_message(self, msg: str) -> str:
+        return f'{self.device_info.name} <{self.device_info.address}>: {msg}'
+
+    def run(self) -> None:
+        t0 = time()
+        try:
+            device = self.connect()
+            ResetDeviceThread.reset_device(device)
+            self.success = self.disconnect_event.wait(self.reset_timeout)
+        except Exception as e:
+            self.logger.exception(e)
+        finally:
+            if self.success:
+                self.logger.debug(self.format_message(f'Reset took {time() - t0:0.1f} sec.'))
+            else:
+                self.logger.debug(self.format_message(f'Reset timed out!'))
+
+    def connect(self) -> MetaWear:
+        """
+        Attempt to connect to the device. Raise an exception if the maximum number of attempts is exceeded.
+
+        :returns: The connected Mbient device object.
+        """
+        device = MetaWear(self.device_info.address)
+
+        success = False
+        self.logger.debug(self.format_message(f'Attempting Connection'))
+        for i in range(self.connect_attempts):
+            try:
+                if i > 0:  # Do not immediately try to reconnect if it just failed.
+                    sleep(3)
+
+                device.connect()
+                success = True
+                break
+            except Exception as e:
+                self.logger.debug(self.format_message(f'Failed to connect on attempt {i + 1}: {e}'))
 
-if __name__ == "__main__":
-    macs = {
-        "Mbient_LH_2": "E8:95:D6:F7:39:D2",
-        "Mbient_RH_2": "FE:07:3E:37:F5:9C",
-        "Mbient_RF_2": "E5:F6:FB:6D:11:8A",
-        "Mbient_LF_2": "DA:B0:96:E4:7F:A3",
-        "Mbient_BK_1": "D7:B0:7E:C2:A1:23",
-        }
-    
-    print('resetting mbients (will take ~ 1 min)...')
-    scann_BLE(2)
-    for k, v in macs.items():
-        success = reset_dev(v)
         if not success:
-            print(f"Failed to connect {k} {v}")
-        else:
-            print(f"Success in resetting {k} {v}")
+            raise Exception(self.format_message(f'Unable to connect!'))
+
+        device.on_disconnect = lambda status: self.on_disconnect(status)
+        return device
+
+    def on_disconnect(self, status) -> None:
+        self.logger.debug(self.format_message(f'Disconnected with status {status}.'))
+        self.disconnect_event.set()
+
+    @staticmethod
+    def reset_device(device: MetaWear) -> None:
```

Comment:
Are we sure this code is thread safe? I didn't research it fully, but saw this in the metaware docs: 

> API Access Queue
The core of this SDK is the MetaWearCpp library, and it’s not indented to be thread safe. Any time you call one of its mbl_mw_* functions you will want to make sure you are on the apiAccessQueue or apiAccessExecutor in order to ensure crash free behavior.

Reply:
I don't see this anywhere in the mbientlab python/c++ docs or examples. (I don't even see the access queue or executor in the API) Maybe this only applies to the smartphone/watch API?

Still need to test in Wang with all sensors. If we see weird issues then definitely worth keeping this in mind!

Reply:
I might be barking up the wrong tree. I found it here: https://mbientlab.com/iosdocs/latest/metawearboard.html